build.rs

```diff
@@ -271,8 +275,78 @@ const MSVC: &'static str = "msvc";
 const MSVC_OBJ_OPT: &'static str = "/Fo";
 const MSVC_OBJ_EXT: &'static str = "obj";
 
+const GENERATED_VERSIONED_HEADER: &str = "generated_versioned.h";
+const GENERATED_YASM_VERSIONED_HEADER: &str = "yasm_versioned.h";
+const GENERATED_VERSIONED_MACRO: &str = "VERSIONED";
+const GENERATED_INCLUDE_DIR: &str = "generated_headers";
+
+fn generated_include_dir() -> PathBuf {
+    PathBuf::from(::std::env::var("OUT_DIR").unwrap()).join(GENERATED_INCLUDE_DIR)
+}
+
+fn custom_versioned_symbols_header() -> PathBuf {
+    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("include").join("versioned.h")
+}
+
+fn yasm_versioned_symbols_header() -> PathBuf {
+    PathBuf::from(::std::env::var("OUT_DIR").unwrap())
+        .join(GENERATED_INCLUDE_DIR)
+        .join(GENERATED_YASM_VERSIONED_HEADER)
+}
+
+// Take the versioned header in `from`, replace #define with %define, preprocess
+// it with the C compiler, write it out to `yasm_versioned_symbols_header()`.
+fn write_yasm_versioned_header(from: &Path) -> io::Result<()> {
+    use std::fs::File;
+    use std::io::{BufReader, BufRead, Write, Cursor};
+
+    // Create the data for the intermediate file.
+    let mut new_data = String::new();
+    for line in BufReader::new(File::open(from)?).lines() {
+        use std::fmt::Write;
+        let line = line?;
+        if line.starts_with("#define") {
+            write!(new_data, "%{}\n", &line[1..]).expect("write to str");
+        } else {
+            write!(new_data, "{}\n", line).expect("write to str");
+        }
+    }
+
+    // Write the data to the intermediate file.
+    let output = yasm_versioned_symbols_header();
+    let intermediate = output.with_extension("in.h");
+    File::create(&intermediate)?.write_all(new_data.as_bytes())?;
+
+    // Expand the intermedia file.
+    let expanded = cc::Build::new()
+        .file(&intermediate)
+        .include(&generated_include_dir())
+        .expand();
```

Comment:
I didn't look at the native_versioning crate's source code yet (since I don't know the license). What is the reason that this has to be run through the preprocessor here?

Reply:
Yes, this whole dance is so that symbols are versioned by `yasm` on Windows as well. We can't use the `versioned.h` header file for `yasm` because its pre-processes uses different syntax. So, instead of having to maintain two `versioned.h` files, one for C compilers and another for `yasm`, I opted to generated a `yasm`-compatible header file in this method from the existing `versioned.h` file.

Reply:
If this is only needed for yasm then we should only run it when we'd run yasm (windows targets).

However, like I mentioned in the other issue, I think that we should get rid of all the `#define`s in versioned.h and instead distribute those `#defines` where the declarations/definitions are. This might result in a small number of cases where we need to duplicate the use of `VERSIONED()` between C and the assembler code, but that would be less problematic than continuously maintaining the versioned.h file as we rename and add symbols.

Reply:
That was my initial approach but quickly proved to be much easier said than done. It's very easy to miss a symbol this way and doing this for the perl generated assembly files is even more difficult. We'd need to modify each generator individually or keep a table of which symbols each generator generates. Then we also have to handle the `yasm` case specially.

I don't think maintaining the `versioned.h` file is any more or less difficult that maintaining `#define`'s in individual files, but I do think that one `versioned.h` file is an overall cleaner solution.

Reply:
> That was my initial approach but quickly proved to be much easier said than done. It's very easy to miss a symbol this way and doing this for the perl generated assembly files is even more difficult.

There are two cases to consider: How difficult it is for you to write this PR, and how difficult it is for us to change  *ring* in the future after this PR. I understand that getting everything right in this PR doing it the way I propose is tricky; however, the upside of doing it the way I propose is that we're less likely to accidentally forget to add/change the `#define` when we add or rename symbols incrementally going forward. In other words, it's going to be somewhat error-prone either way, and we should optimize for minimizing errors in future PRs by people who aren't familiar with this PR.

I think we should go further and just use `GFp_VERSIONED()` in the C declaration and C definition of all of the functions that are only called from Rust code, instead of relying on these `#defines`. We really only would need these `#defines` for functions that are called from the C code, which is smaller and shrinking number of call sites.

For example, instead of:
```c
#define GFp_aes_gcm_open GFp_VERSIONED(GFp_aes_gcm_open)
...
int GFp_aes_gcm_open(const void *ctx_buf, uint8_t *out, size_t in_out_len,
                     uint8_t tag_out[EVP_AEAD_AES_GCM_TAG_LEN],
                     const uint8_t nonce[EVP_AEAD_AES_GCM_NONCE_LEN],
                     const uint8_t *in, const uint8_t *ad, size_t ad_len);
```

We can just have:

```c
int GFp_VERSIONED(GFp_aes_gcm_open)(
                     const void *ctx_buf, uint8_t *out, size_t in_out_len,
                     uint8_t tag_out[EVP_AEAD_AES_GCM_TAG_LEN],
                     const uint8_t nonce[EVP_AEAD_AES_GCM_NONCE_LEN],
                     const uint8_t *in, const uint8_t *ad, size_t ad_len));
````

Because this function is never called from C code.

Reply:
It's not just that what you're proposing is harder to implement, it's that it's a really complex solution that is difficult to get right. I feel confident that what I've implemented here is correct. It's a simple solution that's simple to audit. Were I to manually edit each and every C, assembly, and perl file, I would not be confident that I've covered all bases. I think what you're proposing is a time consuming endeavor with little to no benefit.

In coming up with a solution, maintainability was a big concern of mine. With the present solution, upstream changes from BoringSSL can be merged as easily as before since there are zero changes being made to any of the legacy code. With what you're proposing, each change has to be manually adjusted to properly version symbols.

I imagine most pull requests being made to ring are changes to Rust code; this PR doesn't introduce any overhead there except for using `versioned_extern!` instead of `extern`. But, if a PR is made that changes C/assembly code, I think it's easy for the reviewer to keep in mind that symbols are versioned. Or, at least, no more or less hard than the solution you're advocating for would make it.

I feel like your primary concern is getting this right. That's mine as well. If that's the case, then I propose the following compromise: it's very simple to write a script that checks that all public symbols in object files produced from C/assembly are properly versioned. I'll write such a script, and we'll run it on Travis. This way, it's not possible for PRs to miss versioning symbols. What do you say?